modules/siddhi-extensions/math/src/main/java/org/wso2/siddhi/extension/math/PercentileFunctionExtension.java

```diff
@@ -0,0 +1,393 @@
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * percentile(values, p)
+ * Returns an estimate for the pth percentile of the stored values
+ * Accept Type(s): FLOAT,INT,LONG,DOUBLE
+ * Return Type: DOUBLE
+ */
+public class PercentileFunctionExtension extends AttributeAggregator {
+
+    private PercentileFunctionExtension percentileOutputFunctionExtension;
+    private double percentileValue;
+
+    /**
+     * The initialization method for FunctionExecutor
+     *
+     * @param attributeExpressionExecutors are the executors of each attributes in the function
+     * @param executionPlanContext Execution plan runtime context
+     */
+    @Override
+    protected void init(ExpressionExecutor[] attributeExpressionExecutors, ExecutionPlanContext executionPlanContext) {
+
+        if (attributeExpressionExecutors.length != 2) {
+            throw new OperationNotSupportedException("Percentile function has to have exactly 2 parameter, currently "
+                    + attributeExpressionExecutors.length + " parameters provided");
+        }
+
+        try {
+
+            if (attributeExpressionExecutors[1] instanceof ConstantExpressionExecutor) {
+                percentileValue = ((Double) attributeExpressionExecutors[1].execute(null));
+            } else {
+                throw new OperationNotSupportedException("Percentile value has to be a constant");
+            }
+
+        } catch (ClassCastException c) {
+            throw new ExecutionPlanCreationException("Percentile value should be of type double. But found "
+                    + attributeExpressionExecutors[1].getReturnType());
+        }
+
+        if (percentileValue <= 0 || percentileValue > 100) {
+            throw new OperationNotSupportedException("Percentile value should be in 0 < p ≤ 100 range");
+        }
+
+        Attribute.Type type = attributeExpressionExecutors[0].getReturnType();
+
+        switch (type) {
+        case FLOAT:
+            percentileOutputFunctionExtension = new PercentileFunctionExtensionFloat();
+            break;
+        case INT:
+            percentileOutputFunctionExtension = new PercentileFunctionExtensionInt();
+            break;
+        case LONG:
+            percentileOutputFunctionExtension = new PercentileFunctionExtensionLong();
+            break;
+        case DOUBLE:
+            percentileOutputFunctionExtension = new PercentileFunctionExtensionDouble();
+            break;
+        default:
+            throw new OperationNotSupportedException("Percentile not supported for " + type);
+        }
+    }
+
+    public Attribute.Type getReturnType() {
+        return percentileOutputFunctionExtension.getReturnType();
+    }
+
+    @Override
+    public Object processAdd(Object data) {
+        // will not occur
+        return new IllegalStateException("Percentile cannot process a single argument, but found " + data.toString());
+    }
+
+    @Override
+    public Object processAdd(Object[] data) {
```

Comment:
It seems, caller method is thread safe then no need to synchronized this method again..